### ISSUE TITLE
Lost connection closes window without ability to save query

### ISSUE THREAD

**Issue by jeffkreska** [sequelpro/sequelpro#2838]
Might not be a bug but it would be a nice enhancement if it isnt a bug to allow the sql in the tab of a connection that was lost to be copied or saved before destroying the tab.

Steps to reproduce:
1. Run some sql in the query editor
2. Let sit until connection times out (disconnect network to box)
3. Dialog is displayed saying somthing like "Connection lost" Reconnect or disconnect?
4. Click Reconnect, but it fails
5. Dialog is displayed again

Expected the when the dialog prompting to reconnect is displayed there is an option to save sql in query editor before the tab is closed.

**Comment by celdaran** [comment]
+1000 on this one. I've been amazed that this hasn't: 1) been reported more or 2) been fixed already. I've come to expect that one fundamental aspect of any application is "don't indiscriminately discard the user's work."

While I realize that I'm partially to blame for not saving large query buffers more frequently, there are times when in the heat of battle, it just slips my mind and more than once Sequel Pro has assumed "can't connect to the database" is equivalent to "throw away all the work."

**Comment by Dunnzilla** [comment]
"Don't indiscriminately discard the user's work" about sums it up.

**Comment by neurogenesis** [comment]
I think this should be classified as a bug. Since the "reconnect" dialog is a modal window, it essentially places the application in an inaccessible / unusable state until the tab / window is closed. Even if the full unsaved query/session was visible in the window, you are unable to open a new connection / session from the application menu.

If the application is unable to reconnect, the expected behavior is that hitting escape or cancel on the dialog box should return the user to the editor. The current query should be cancelled, while future queries that fail to connect should again correctly prompt the user with the reconnect dialog.

Currently, hitting escape or otherwise closing the dialog will close the session tab and result in loss of work.

**Comment by neurogenesis** [comment]
@dmoagx : any thoughts on reconnect logic that might allow a user to suspend retries long enough to save their work before the tab is closed?

**Comment by dmoagx** [comment]
@neurogenesis That reconnect dialog isn't there because we like it, but because the rest of the application simply cannot handle an open connection tab without an active connection. Fixing that requires some large scale rearchitecturing.

**Comment by celdaran** [comment]
@dmoagx since "some large scale rearchitecturing" probably isn't on the immediate bug fix list, I'd be happy with some sort of auto-save feature. Set a timer. Every five or ten minutes, just click File|Save for me. That'd be sweet too.  :)

**Comment by nikoskip** [comment]
+1

An auto-save mode will work or maybe a new option like "Close connection and save queries".

**Comment by doughertym** [comment]
+1

**Comment by brendan-roche** [comment]
I just got bitten by this one again, and lost a bunch of queries I've been using, very frustrating!
I've been using this Sequel Pro for some time and love the simplicity of the tool, but this is the only issue that bugs me more than anything else. 

Whenever i encounter this issue, "Reconnect" has never worked for me. However I am able to successfully open the same connection in another tab with no problems, so I'm not sure why i can never reconnect. 
  
@dmoagx, if you say to get reconnecting working will require some large re-architecturing, there are quite a few simpler options to avoid users losing their unsaved work, which is really the main issue here.

I agree with @neurogenesis that a simple "Cancel" option should be present, this will cancel the running query and allow a user to resume and save any queries or results they have in the connection tab. Additionally (or alternatively) when the option to "close" is chosen the user would see an additional dialog indicating they have unsaved queries and would they like to save it.

However the best solution would be to just save users session as they make changes. So closing and reopening the application all my connection tabs would be reopened with my saved sessions so I'm able to immediately pick up from where i was working.

@dmoagx it would be really great if this could be prioritised as it gives user no recourse to resume on failure, thereby losing a whole lot of time in having to rewrite queries. I can't tell you how much time I've lost in the past on this. And it's true, given that it does happen from time to time I should be saving more often, but pretty much all other developer tools I've used will save the active session, but I'd be happy for any solution right now.

Thanks for providing a fantastic tool btw :)
